torchci/lib/bot/greenBot.ts

```diff
@@ -0,0 +1,231 @@
+
+const openOnGreen = new RegExp("^s*@pytorchbot\\s+?mergeOnGreen.*");
+export const LAND_PENDING = "land-pending";
+const LAND_ATTEMPT = "landed-attempt";
+const LAND_FAILED = "land-failed";
+
+type OctokitData = {
+  octokit: InstanceType<typeof ProbotOctokit>;
+  owner: string;
+  repo: string;
+  issue_number: number;
+  comment: string;
+};
+
+export default function greenBot(app: Probot): void {
+  app.on("issue_comment.created", async (ctx) => {
+    const commentBody = ctx.payload.comment.body;
+    if (commentBody.match(openOnGreen)) {
+      if (!ctx.payload.issue.pull_request) {
+        // Issue, not pull request.
+        await reactOnComment(ctx, "confused");
+      } else {
+        await reactOnComment(ctx, "+1");
+        await addLabels(ctx, [LAND_PENDING]);
+      }
+    }
+  });
+
+  app.on(["pull_request.labeled"], async (ctx) => {
+    if (ctx.payload.label.name === LAND_PENDING) {
+      // Check pull request's head ref and check runs to see if it has all green and merge if it's good
+      // TODO: Add pagination if checks go above 100
+      const payloadData = getPayloadData(ctx);
+      const ref = ctx.payload.pull_request.head.sha;
+      const prNum = ctx.payload.pull_request.number;
+      const jobs = await getJobsBySha(ref);
+      const [successfulJobs, failedJobs] = categorizeJobs(jobs);
+      // If all jobs are successful or neutral, then try merge, else do nothing
+      if (successfulJobs.length === jobs.length) {
+        await tryLandPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "All checks passed. Attempting to merge.",
+        });
+      }
+      // If any failed, alert the author
+      else if (failedJobs.length > 0) {
+        await failPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "Failed to land PR due to red signal.",
+        });
+      }
+    }
+  });
+
+  app.on("workflow_job.completed", async (ctx) => {
+    const rocksetClient = getRocksetClient();
+    // Get all PRs that have this as the head SHA and the land pending label
+    const landPendingPrQuery =
+      await rocksetClient.queryLambdas.executeQueryLambda(
+        "commons",
+        "prs_with_label",
+        rocksetVersions.commons.prs_with_label,
+        {
+          parameters: [
+            {
+              name: "label",
+              type: "string",
+              value: LAND_PENDING,
+            },
+            {
+              name: "sha",
+              type: "string",
+              value: ctx.payload.workflow_job.head_sha,
+            },
+          ],
+        }
+      );
+
+    const prs = landPendingPrQuery?.results;
+    if (prs == null || prs?.length === 0) {
+      return;
+    }
+
+    if (prs.length > 1) {
+      // If there's more than 1 PR with the same sha and the land-pending label,
+      // Then we don't want to potentially race.
+      return;
```

Comment:
Is returning the best behavior here though? Will we notify the user somehow with actionable next steps?